clash-cores/src/Clash/Cores/Xilinx/Ila/Internal.hs

```diff
@@ -0,0 +1,360 @@
+-- XXX: I'd move this 'deriveTermLiteral' up, but Template Haskell complains..
+deriveTermLiteral ''ProbeType
+instance KnownNat n => TermLiteral (IlaConfig n) where
+  termToData = $(deriveTermToData ''IlaConfig)
+
+probeTypesVec :: KnownNat n => IlaConfig n -> Vec n ProbeType
+probeTypesVec = either C.repeat id . probeTypes
+
+comparatorsVec :: KnownNat n => IlaConfig n -> Vec n Int
+comparatorsVec = either C.repeat id . comparators
+
+areSame :: Eq a => [a] -> Maybe a
+areSame = \case
+  c:cs -> foldl go (Just c) cs
+  [] -> Nothing
+ where
+  go Nothing _ = Nothing
+  go (Just c0) c1
+    | c0 == c1 = Just c0
+    | otherwise = Nothing
+
+ilaBBF :: HasCallStack => BlackBoxFunction
+ilaBBF _isD _primName args _resTys
+  | _:_:config:_ <- lefts args
+  , _:_:LitTy (NumTy n):_ <- rights args
+  , Just (SomeNat (Proxy :: Proxy n)) <- someNatVal n
+  = case termToDataError @(IlaConfig n) config of
+      Left s -> error ("ilaBBF, bad config:\n" <> s)
+      Right c -> pure $ Right (bbMeta c, bb c)
+  | otherwise = error $ "ilaBBF, bad args:\n" <> ppShow args
+ where
+   bbMeta :: KnownNat n => IlaConfig n -> BlackBoxMeta
+   bbMeta config = emptyBlackBoxMeta
+     { bbKind = TDecl
+     , bbRenderVoid = RenderVoid
+     , bbIncludes =
+         [ ( ("ila", "clash.tcl")
+           , BBFunction (show 'ilaTclTF) 0 (ilaTclTF config)
+           )
+         ]
+     }
+
+   bb :: KnownNat n => IlaConfig n -> BlackBox
+   bb config = BBFunction (show 'ilaTF) 0 (ilaTF config)
+
+usedArguments :: [Int]
+usedArguments = (ilaConfig : clock : inputProbes)
+ where
+  (    _knownDomain
+    :< _ilaConstraint
+    :< ilaConfig
+    :< clock
+    :< (Infinite.take 8096 -> inputProbes)
+    ) = ((0::Int)...) -- This function is polyvariadic so in theory it supports an
+                      -- unlimited number of arguments. To prevent evaluation loops
+                      -- when forcing this argument to NF we limit it to a modest
+                      -- 8096 input ports.
+
+ilaTF :: (HasCallStack, KnownNat n) => IlaConfig n -> TemplateFunction
+ilaTF config =
+  TemplateFunction
+    usedArguments
+    -- 'validateIlaProbeBCC' already produces string describing
+    -- failing checks, but 'TemplateFunction' cannot handle this
+    -- yet. This is prepared to get updated easily as soon as the
+    -- feature gets implemented in clash-lib.
+    (maybe True error . validateIlaProbeBBC)
+    (ilaBBTF config)
+
+checkNameCollision :: HasCallStack => T.Text -> DSL.TExpr -> DSL.TExpr
+checkNameCollision userName tExpr@(DSL.TExpr _ (Identifier (Id.toText -> name) Nothing))
+  | userName == name = tExpr
+  | otherwise = error [__i|
+      Tried create a signal called '#{userName}', but identifier generation returned
+      '#{name}'. Refusing to instantiate Ila with unreliable probe names.
+  |]
+checkNameCollision _ tExpr = error [__i|
+  Internal error: Expected 'TExpr' with the following form:
+
+    TExpr _ (Identifier _ Nothing)
+
+  got:
+
+    #{ppShow tExpr}
```

Comment:
I don't think you should indent your `ppShow`. I'm almost certain this will indent the first line only, having subsequent lines outdent by two spaces again, only making the output more difficult to read.

This happens in multiple places, I'm just commenting once.

Reply:
Hmm, the quasiquoter I wrote many moons ago handled this correctly. But it's drowned out by the 1001 other QQs so I'm not sure what I'm using now :upside_down_face:

Reply:
`I.i` handles this :)